### PR TITLE
HAI-2342: Producing tormays_cycle_infra_polys

### PR DESCRIPTION
**Pull request by petri-timlin**
### Description

- Added cycle_infra to automation
- Added own buffer setting for cycle_infra ylre_street_classses
- Changed db connection to use also env variables
- Added field "kadun_nimi" for ylre_katualueet
- Added field "kadun_nimi" for cycle_infra data. This data is from ylre_katualueet.
- Added geometry validation for cycle_infra data
- Cycle_infra data is consist of both by hierarchy and by paatyyppi/alatyyppi

### Jira Issue:
https://helsinkisolutionoffice.atlassian.net/browse/HAI-2342

### Type of change

- [ ] Bug fix
- [X] New feature
- [ ] Other

### Instructions for testing
Instructions can be found from: automation/README.md:

Give following in folder /automation:
`docker build -t haitaton-gis-automation -f ./Dockerfile ..`
This wil build haitaton-gis-automation image.

Run image
Give following:
`docker run --rm --network host --env-file haitaton.env haitaton-gis-automation cycle_infra`

After this Cycle infra data is written to database into the table `tormays_cycle_infra_polys`
and there should be following fields: 
- fid
- paatyyppi
- alatyyppi
- silta_alikulku
- hierarkia
- yksisuuntaisuus
- kadun_nimi
- geom

In local environment there might be situation that table `tormays_cycle_infra_polys` is missing but after docker run table should exists.

### CODE REVIEW

File: automation/README.md [hunk 1]
```diff
@@ -44,10 +44,8 @@ Where `<source>` is currently one of:
 - `ylre_katualueet` - Helsinki YLRE street areas, polygons.
 - `ylre_katuosat` - Helsinki YLRE parts, polygons.
 - `maka_autoliikennemaarat` - Traffic volumes (car traffic)
-- ~~`cycle_infra` - Cycle infra (local file)~~
+- `cycle_infra` - Cycle infra (local file)
```
Comment (corvidian) [inline automation/README.md]: This is no longer a local file, right?

File: process/modules/ylre_katualueet.py [hunk 2]
```diff
@@ -36,7 +36,7 @@ def purpose_to_class(purpose: str) -> str:
         )
         df = df[~df["ylre_class"].isna()]
         df = df[~(df["geometry"].isna() | df["geometry"].is_empty)].loc[
-            :, ["ylre_class", "geometry"]
+            :, ["ylre_class", "kadun_nimi", "geometry"]
```
Comment (corvidian) [inline process/modules/ylre_katualueet.py:39]: Is this related to HAI-2416?

Comment (petri-timlin) [reply process/modules/ylre_katualueet.py:39]: No. Need to add "kadun_nimi" attribute for ylre_katualueet because it is used for to get that information for cycle_infra.

File: automation/README.md (hunk 1 above)
Comment (petri-timlin) [reply automation/README.md]: Changed.

File: process/modules/cycling_infra.py [hunk 3]
```diff
@@ -18,22 +26,221 @@ def __init__(self, cfg: Config):
...
+        target_infra_polys.rename(columns={"index_right1": "kadun_nimi"}, inplace=True)
+        # Drop unnecessary columns
+        target_infra_polys = self._drop_unnecessary_columns(
+            self._dropped_columns, target_infra_polys
+        )
+
+        # Dissolve geometries base on attributes listed
+        attrs = ["paatyyppi", "alatyyppi", "silta_alikulku", "hierarkia", "yksisuuntaisuus", "kadun_nimi", ]
+        for attr in attrs:
+            target_infra_polys[attr] = target_infra_polys[attr].fillna("NULL")
```
Comment (corvidian) [inline process/modules/cycling_infra.py]: The `tormays_cycle_infra_polys` table contains "NULL" strings instead of null values. Perhaps because of this line?

```
haitaton=# select distinct yksisuuntaisuus from tormays_cycle_infra_polys ;
                yksisuuntaisuus
------------------------------------------------
 Yksisuuntainen digitointisuuntaan
 Kaksisuuntainen
 Yksisuuntainen (ei tietoa digitointisuunnasta)
 NULL
 Yksisuuntainen digitointisuuntaa vastaan
(5 rows)

haitaton=# update tormays_cycle_infra_polys set yksisuuntaisuus = NULL where yksisuuntaisuus='NULL';
UPDATE 786

haitaton=# select distinct yksisuuntaisuus from tormays_cycle_infra_polys ;
                yksisuuntaisuus
------------------------------------------------
 Yksisuuntainen (ei tietoa digitointisuunnasta)
 Yksisuuntainen digitointisuuntaa vastaan
 <null>
 Yksisuuntainen digitointisuuntaan
 Kaksisuuntainen
(5 rows)
```

**Comment by corvidian** [comment]
image
It's been said that the cycle path network in Helsinki is sparse, but it's not this sparse. The data is missing a lot of the big cycle paths around where I live because they're categorized as `Puistotie- tai väylä` in the data.

File: fetch/fetch_data.sh [hunk 4]
```diff
@@ -65,15 +65,15 @@ hsl)
     wget -O "$local_file" "$addr"
     ;;
 # plain cp from local file system
-cycle_infra)
+cycle_infra_old)
```
Comment (corvidian) [inline fetch/fetch_data.sh]: Do we need the old fetch for something?

File: fetch/fetch_data.sh [hunk 5]
```diff
@@ -65,15 +65,15 @@ hsl)
...
     cp "$addr" "$local_file"
     ;;
 # plain WFS fetch
 hki|ylre_katualueet|ylre_katuosat|maka_autoliikennemaarat|osm|helsinki_osm_lines|central_business_area)
     ogr2ogr -progress -f GPKG "$local_file" ${extra_args:+$extra_args} ${extra_quoted_args:+"$extra_quoted_args"} "$addr" "$layer"
     ;;
 # plain WFS fetch with authentication
-liikennevaylat)
+liikennevaylat|cycle_infra)
```
Comment (corvidian) [inline fetch/fetch_data.sh:72]: Root folder README has outdated info on what the cycle_infra task does and needs.

File: fetch/fetch_data.sh (hunk 4 above)
Comment (petri-timlin) [reply fetch/fetch_data.sh]: No, I forgot to remove it.

File: fetch/fetch_data.sh (hunk 5 above)
Comment (petri-timlin) [reply fetch/fetch_data.sh:72]: Fixed

File: process/modules/cycling_infra.py (hunk 3 above)
Comment (petri-timlin) [reply process/modules/cycling_infra.py]: Fixed

File: fetch/fetch_data.sh (hunk 4 above)
Comment (petri-timlin) [reply fetch/fetch_data.sh]: Fixed

**Comment by petri-timlin** [comment]
Added also "Puistotie- tai väylä" material to cycle_infra.

**Comment by corvidian** [comment]
> Added also "Puistotie- tai väylä" material to cycle_infra.

This looks better now.
image

**Review by corvidian — APPROVED** [review]
